### PR TITLE
Replace `CConstruction::Frames` from intrusive list to `std::vector<C…

### PR DESCRIPTION
**Pull request by Jarod42**
…ConstructionFrame>`

### CODE REVIEW

File: src/stratagus/construct.cpp [hunk 1]
```diff
@@ -264,15 +258,10 @@ static int CclDefineConstruction(lua_State *l)
...
-				CConstructionFrame **cframe = &construction->Frames;
-				while (*cframe) {
-					cframe = &((*cframe)->Next);
-				}
-				(*cframe) = new CConstructionFrame;
-				(*cframe)->Percent = percent;
-				(*cframe)->File = file;
-				(*cframe)->Frame = frame;
-				(*cframe)->Next = nullptr;
+				auto& cframe = construction->Frames.emplace_back();
```
Comment (zzam) [inline src/stratagus/construct.cpp:261]: Could the size be set before to avoid re-allocation?

File: src/action/action_built.cpp [hunk 2]
```diff
@@ -331,17 +321,12 @@ void COrder_Built::AiUnitKilled(CUnit &unit)
 }
 
 
-static const CConstructionFrame *FindCFramePercent(const CConstructionFrame &cframe, int percent)
+static std::size_t FindCFramePercent(const std::vector<CConstructionFrame> &cframes, int percent)
 {
-	const CConstructionFrame *prev = &cframe;
+	const auto it =
+		ranges::find_if(cframes, [&](const auto &frame) { return percent < frame.Percent; });
```
Comment (zzam) [inline src/action/action_built.cpp:327]: Is it worth doing binary_search here?
And is it even guaranteed/checked that that Percent-values are ordered correctly?

Comment (Jarod42) [reply src/action/action_built.cpp:327]: I tested on wargus and it seems fine...
I will recheck.

File: src/stratagus/construct.cpp (hunk 1 above)
Comment (Jarod42) [reply src/stratagus/construct.cpp:261]: https://github.com/Wargus/stratagus/pull/637 created.

File: src/action/action_built.cpp (hunk 2 above)
Comment (Jarod42) [reply src/action/action_built.cpp:327]: Well spot, I indeed have bad offset.
I also add pre-require check in https://github.com/Wargus/stratagus/pull/638.

**Pull request merged**